src/ast/ast_pp_dot.cpp

```diff
@@ -0,0 +1,132 @@
+/*++
+
+Abstract: Pretty-printer for proofs in Graphviz format
+
+--*/
+
+#include "util/util.h"
+#include "util/map.h"
+#include "ast_pp_dot.h"
+
+// string escaping for DOT
+std::string escape_dot(std::string const & s) {
+    std::string res;
+    for (auto c : s) {
+        if (c == '\n')
+            res.append("\\l");
+        else
+            res.push_back(c);
+    }
+    return res;
+}
+
+// map from proofs to unique IDs
```

Comment:
We typically use:
obj_map<expr, unsigned> and 
obj_map<expr, bool> 

It may not make a visible difference here, but generally pointers used for comparisons in maps make their behavior non-reproducible. The obj_map uses unique identifiers associated with the first arguments.

Reply:
Ok, will do. I was wondering what kind of value to use for the set (I'd use `unit` in ML, but I don't know of any equivalent in C++ so I used `bool`)?

Reply:
Use obj_hashtable.